frame/offences/src/lib.rs

```diff
@@ -57,59 +57,86 @@ impl WeightInfo for () {
-	/// Full identification of the validator.
-	type IdentificationTuple: Parameter + Ord;
-	/// A handler called for every offence report.
-	type OnOffenceHandler: OnOffenceHandler<Self::AccountId, Self::IdentificationTuple, Weight>;
-}
-
-decl_storage! {
-	trait Store for Module<T: Config> as Offences {
-		/// The primary structure that holds all offence records keyed by report identifiers.
-		Reports get(fn reports):
-			map hasher(twox_64_concat) ReportIdOf<T>
-			=> Option<OffenceDetails<T::AccountId, T::IdentificationTuple>>;
-
-		/// A vector of reports of the same kind that happened at the same time slot.
-		ConcurrentReportsIndex:
-			double_map hasher(twox_64_concat) Kind, hasher(twox_64_concat) OpaqueTimeSlot
-			=> Vec<ReportIdOf<T>>;
-
-		/// Enumerates all reports of a kind along with the time they happened.
-		///
-		/// All reports are sorted by the time of offence.
-		///
-		/// Note that the actual type of this mapping is `Vec<u8>`, this is because values of
-		/// different types are not supported at the moment so we are doing the manual serialization.
-		ReportsByKindIndex: map hasher(twox_64_concat) Kind => Vec<u8>; // (O::TimeSlot, ReportIdOf<T>)
+#[frame_support::pallet]
+pub mod pallet {
+	use super::*;
+	use frame_support::pallet_prelude::*;
+	use frame_system::pallet_prelude::*;
+
+	#[pallet::pallet]
+	#[pallet::generate_store(pub(super) trait Store)]
+	pub struct Pallet<T>(_);
+
+	/// The pallet's config trait.
+	#[pallet::config]
+	pub trait Config: frame_system::Config {
+		/// The overarching event type.
+		type Event: From<Event> + IsType<<Self as frame_system::Config>::Event>;
+		/// Full identification of the validator.
+		type IdentificationTuple: Parameter + Ord;
+		/// A handler called for every offence report.
+		type OnOffenceHandler: OnOffenceHandler<Self::AccountId, Self::IdentificationTuple, Weight>;
 	}
-}
 
-decl_event!(
+	/// The primary structure that holds all offence records keyed by report identifiers.
+	#[pallet::storage]
+	#[pallet::getter(fn reports)]
+	pub type Reports<T: Config> = StorageMap<
+		_,
+		Twox64Concat,
+		ReportIdOf<T>,
+		OffenceDetails<T::AccountId, T::IdentificationTuple>,
+	>;
+
+	/// A vector of reports of the same kind that happened at the same time slot.
+	#[pallet::storage]
+	pub type ConcurrentReportsIndex<T: Config> = StorageDoubleMap<
+		_,
+		Twox64Concat,
+		Kind,
+		Twox64Concat,
+		OpaqueTimeSlot,
+		Vec<ReportIdOf<T>>,
+		ValueQuery,
+	>;
+
+	/// Enumerates all reports of a kind along with the time they happened.
+	///
+	/// All reports are sorted by the time of offence.
+	///
+	/// Note that the actual type of this mapping is `Vec<u8>`, this is because values of
+	/// different types are not supported at the moment so we are doing the manual serialization.
+	#[pallet::storage]
+	pub type ReportsByKindIndex<T> = StorageMap<
+		_,
+		Twox64Concat,
+		Kind,
+		Vec<u8>, // (O::TimeSlot, ReportIdOf<T>)
```

Comment:
In order not to change the storage, it must be written with a ValueQuery
```suggestion
		Vec<u8>, // (O::TimeSlot, ReportIdOf<T>)
		ValueQuery,
```